test/unit/mem/buf/src/BufTest.cpp

```diff
@@ -41,6 +41,35 @@ static auto testBufferMutable(alpaka::Vec<alpaka::Dim<TAcc>, alpaka::Idx<TAcc>>
     alpaka::test::testViewMutable<TAcc>(queue, buf);
 }
 
+template<typename TAcc>
+static auto testAsyncBufferMutable(alpaka::Vec<alpaka::Dim<TAcc>, alpaka::Idx<TAcc>> const& extent) -> void
+{
+    using Dev = alpaka::Dev<TAcc>;
+    using Pltf = alpaka::Pltf<Dev>;
+    using Queue = alpaka::test::DefaultQueue<Dev>;
+
+    using Elem = float;
+    using Dim = alpaka::Dim<TAcc>;
+    using Idx = alpaka::Idx<TAcc>;
+
+    Dev const dev = alpaka::getDevByIdx<Pltf>(0u);
+    Queue queue(dev);
+
+    // memory is allocated when the queue reaches this point
+    auto buf = alpaka::allocAsyncBuf<Elem, Idx>(queue, extent);
+
+    // asynchronous operations can be submitted to the queue immediately
+    alpaka::test::testViewMutable<TAcc>(queue, buf);
+
+    // synchronous operations must wait for the memory to be available
+    alpaka::wait(queue);
+    auto const offset = alpaka::Vec<Dim, Idx>::zeros();
+    alpaka::test::testViewImmutable<Elem>(buf, dev, extent, offset);
+
+    // memory will be deallocated once the queue reaches this point
+    alpaka::wait(queue);
+}
+
```

Comment:
@bernhardmgruber did you have in mind something like this, or something else ?

Given that queue-ordered memory allocations are available only for CUDA and CPU buffers, I expect that the tests will fail for the other accelerators.
What would be a good way to make the new test run conditionally on the accelerator type ?

Also, is there a way to run the tests locally ?
I couldn't find any instructions for it.

Reply:
Something of that size looks fine to me for starters, thank you! I think you could add more tests on the lifetime aspect of these async buffers, like enqueuing a kernel, then destroying the buffer instance, then waiting for the kernel that still uses that buffer.

> What would be a good way to make the new test run conditionally on the accelerator type ?

Template specialization or overloading. I would pick the latter, add the `Acc` as unused first argument, then add empty overloads of `testAsyncBufferMutable` for the specific accelerators where the test does not work yet.

> Also, is there a way to run the tests locally ?

alpaka uses cmake's standard testing facilities. Set `BUILD_TESTING` to `ON` in cmake and build all the test targets (or just all targets since the examples should compile as well, in case you turned them on). Run `ctest` in the build dir to execute all tests.